### ISSUE TITLE
all-zero gradients for temp_norm and old_temp

### ISSUE THREAD

**Issue by gsp-leon** [JiangpengHe/imbalanced_cil#5]
Hello, thank you for open-sourcing this code. I'm very interested in the methodology presented in your paper and have been attempting to reproduce the approach. 
Using the CIFAR100-LT dataset with LFS and a total of 10 tasks, I encountered an issue during Task 1. In DGR.py, the line:
temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0]
produces ​all-zero gradients for temp_norm, despite both loss_cn and ele.weight being non-zero variables. What might be causing this issue?
Looking forward to your reply. Thanks.

**Comment by JiangpengHe** [comment]
Thanks for your interest and sorry for the late response, the line "temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0]" is used starting from the second task (Task 2). The gradient for first task (Task 1)_ is calculated in the line 193-209 in DGR.py. Please let me know if this addresses your issue and any other questions. Thanks.

**Comment by gsp-leon** [comment]
Thank you for your reply. It is possible that I have expressed myself incorrectly.
My first task is task0, its gradient is calculated in DGR.py lines 193-209. 
Task1 is my second task that produces the problem： temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0]produces all-zero gradients for temp_norm, despite both loss_cn and ele.weight being non-zero variables.
In order to clearly express my problem, I've attached a screenshot of my resultant output. As a demonstration, I used the CIFAR100-LT dataset with LFS and a total of 10 tasks,nepochs=1. As shown in the figure, since old_temp = 0 and old_temp is the denominator, this results in loss = nan for subsequent operations
Looking forward to your reply. Thanks.
![Image](https://github.com/user-attachments/assets/1f0cc59a-831c-4157-af29-5b280eb2a82b)

**Comment by JiangpengHe** [comment]
Thanks for the clarification, if temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0] is all 0 usually means the loss_cn is calculated without depending on the ele, can you try "loss_cn.backward(retain_graph=True)
print(ele.weight.grad)" to see if gradient is None or not? Also, can you try with a larger epoch instead of 1?

**Comment by gsp-leon** [comment]
Thank you for your suggestion. I modified the code as you suggested:
for ele in self.model.heads:
                    loss_cn.backward(retain_graph=True)
                    print(ele.weight.grad)
                    temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0] 
In the for loop, the first time print(ele.weight.grad) is a non-zero variable and temp_norm is all non-zero. The second time print(ele.weight.grad) is a non-zero variable, but temp_norm is all zeros.
Secondly, about the epoch adjustment, I change epoch=160, still have the same problem, the loss=nan in each epoch, so I don't think the number of epochs is the cause of the problem.

**Comment by gsp-leon** [comment]
Thank you for your suggestion. I modified the code as you suggested 
for ele in self.model.heads:
                    loss_cn.backward(retain_graph=True)
                    print(ele.weight.grad)
                    temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0] 

In the for loop, the first time print(ele.weight.grad) is a non-zero variable and temp_norm is all non-zero. The second time print(ele.weight.grad) is a non-zero variable, but temp_norm is all zeros.
Secondly, about the epoch adjustment, I change epoch=160, still have the same problem, the loss=nan in each epoch, so I don't think the number of epochs is the cause of the problem.

**Comment by gsp-leon** [comment]
Thank you for your suggestion. I modified the code as you suggested

```
for ele in self.model.heads:
       loss_cn.backward(retain_graph=True)
       print(ele.weight.grad)
       temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0]

```

In the for loop, the first time print(ele.weight.grad) is a non-zero variable and temp_norm is all non-zero. The second time print(ele.weight.grad) is a non-zero variable, but temp_norm is all zeros.
Secondly, about the epoch adjustment, I change epoch=160, still have the same problem, the loss=nan in each epoch, so I don't think the number of epochs is the cause of the problem.

**Comment by gsp-leon** [comment]
Thank you for your suggestion. I modified the code as you suggested

```
for ele in self.model.heads:
      loss_cn.backward(retain_graph=True)
      print(ele.weight.grad)
      temp_norm = torch.autograd.grad(loss_cn, ele.weight, retain_graph=True)[0]
```

In the for loop, the first time print(ele.weight.grad) is a non-zero variable and temp_norm is all non-zero. The second time print(ele.weight.grad) is a non-zero variable, but temp_norm is all zeros.
Secondly, about the epoch adjustment, I change epoch=160, still have the same problem, the loss=nan in each epoch, so I don't think the number of epochs is the cause of the problem.

**Comment by JiangpengHe** [comment]
Thanks for your update, it’s likely a runtime behavior based on either the batch content or how the graph is being used. You can check whether the current batch has any new-task samples and ensure that autograd.grad() is called before backward() or after a fresh forward pass.

**Comment by gsp-leon** [comment]
Thank you for the reminder. It was my oversight to not focus on the example sample size setting. For the dataset cifar100, the code runs successfully after setting num-exemplars-per-class=20. But I would also like to ask you how to set the number of example samples for cifar100_lt?

**Comment by JiangpengHe** [comment]
Great! For all experiments, we use a memory budget of 20 exemplars per class for simplicity. Therefore, in the long-tailed case, the exemplar set could also be imbalanced since some tail classes contain less than 20 samples.
